dodiscover/scm/causica.py

```diff
@@ -0,0 +1,153 @@
+"""
+Wrapper for the Causical algorithm for causal discovery with a non-linear
+additve SCM.
+"""
+from dodiscover.context import Context
+
+from typing import List, Literal, NamedTuple, Optional, Tuple, Union
+
+import networkx as nx
+import numpy as np
+import pandas as pd
+import torch
+
+import tempfile
+
+torch.set_default_dtype(torch.float32)
+
+
+class DefaultModelOptions(NamedTuple):
+    base_distribution_type: Literal["gaussian", "spline"] = "spline"
+    spline_bins: int = 8
+    imputation: bool = False
+    lambda_dag: float = 100.0
+    lambda_sparse: float = 5.0
+    tau_gumbel: float = 1.0
+    var_dist_A_mode: Literal["simple", "enco", "true", "three"] = "three"
+    imputer_layer_sizes: Optional[List[int]] = None
+    mode_adjacency: Literal["upper", "lower", "learn"] = "learn"
+    norm_layers: bool = True
+    res_connection: bool = True
+    encoder_layer_sizes: Optional[List[int]] = [32, 32]
+    decoder_layer_sizes: Optional[List[int]] = [32, 32]
+    cate_rff_n_features: int = 3000
+    cate_rff_lengthscale: Union[int, float, List[float], Tuple[float, float]] = 1
+
+
+class DeciTrainingOptions(NamedTuple):
+    learning_rate: float = 3e-2
+    batch_size: int = 512
+    standardize_data_mean: bool = False
+    standardize_data_std: bool = False
+    rho: float = 10.0
+    safety_rho: float = 1e13
+    alpha: float = 0.0
+    safety_alpha: float = 1e13
+    tol_dag: float = 1e-3
+    progress_rate: float = 0.25
+    max_steps_auglag: int = 20
+    max_auglag_inner_epochs: int = 1000
+    max_p_train_dropout: float = 0.25
+    reconstruction_loss_factor: float = 1.0
+    anneal_entropy: Literal["linear", "noanneal"] = "noanneal"
+    device: Literal["cpu", "gpu"] = "cpu"
+
+
+class DECI:
+    def __init__(self, model_params: dict):
+        full_model_options = DefaultModelOptions()._asdict()
+        full_model_options.update(model_params)
+        self.full_model_options = full_model_options
+
+    def fit(
+        self,
+        data: pd.DataFrame,
+        context: Context,
+        training_options: dict,
```

Comment:
I'm in favor of keeping fit API as `fit(data, context)` and any model options within the initialization. Keeps consistent across the different models